### PR TITLE
eth_sign update

### LINKED ISSUE #1531: Switch to eth_sign-compatible signatures in core and SCs
## Problem Definition

We currently sign messages on Raiden simply by signing the hash of the message. By using [eth_sign_sha3](https://github.com/raiden-network/raiden/blob/2df06259496fd10cadd29c8e3f4f5f86289f3940/raiden/utils/__init__.py#L57-L64) as default hasher  in `signing.sign`, and changing the smartcontracts accordingly (should be as simple as prepending the prefix before `keccak256` it), we can make all our signing compatible with `eth_sign` RPC interface.

Of course, it doesn't mean we fully support this standard, but as the changes are minimal, it could allow us to fully use a web3 signer middleware, and **possibly** add in the future the option of doing all signing on the eth-node side, getting rid of opening the private key in the Raiden node, and easily switching to eip712 signatures when finished.

Related: #1293 #1402 #1467 #1475

### LINKED ISSUE #2240: Remove partner argument from setTotalWithdraw.
# Problem Definition

The `partner` argument for `setTotalWithdraw` was [removed](https://github.com/raiden-network/raiden-contracts/pull/251), this needs to be done in the client as well.

### LINKED ISSUE #2325: Use a eth_sign and EIP191 compatible signature format
Related:
https://github.com/raiden-network/raiden/issues/1531

Implemented in the smart contracts here: https://github.com/raiden-network/raiden-contracts/pull/252

Compatibility with EIP191 & `eth_sign`:

```
\x19Ethereum Signed Message:\n" + len(message) + message
message = <token_network_address> <chain_id> <message_type_id> <message specific data to sign>
```

## Implementation in the smart contracts

```
string public constant signature_prefix = '\x19Ethereum Signed Message:\n';
```

Message content is tightly packed as described here: https://solidity.readthedocs.io/en/v0.4.24/abi-spec.html#abi-packed-mode
Python tests implementation: https://github.com/raiden-network/raiden-contracts/pull/252/commits/1569bb635b178c9683c9714b0c92dff4899b4afd#diff-6180c16583e73c52c0366184d36015f9 (I would have used `encode_abi_packed` from https://eth-abi.readthedocs.io/en/latest/encoding.html#non-standard-packed-mode-encoding, but there are some dependencies issues at the moment)

### Balance Proof Message:

```js
        uint256 message_type_id = 1;

        // Length of the actual message: 20 + 32 + 32 + 32 + 32 + 32 + 32
        string memory message_length = '212';

        bytes32 message_hash = keccak256(abi.encodePacked(
            signature_prefix,  // string
            message_length,  // string
            token_network_address,  // address
            chain_id,  // uint256
            message_type_id,  // uint256
            channel_identifier,  // uint256
            balance_hash,  // bytes32
            nonce,  // uint256
            additional_hash  // bytes32
        ));
```

### Balance Proof Update Message:

```
        uint256 message_type_id = 2;

        // Length of the actual message: 20 + 32 + 32 + 32 + 32 + 32 + 32 + 65
        string memory message_length = '277';

        bytes32 message_hash = keccak256(abi.encodePacked(
            signature_prefix,  // string
            message_length,  // string
            token_network_address,  // address
            chain_id,  // uint256
            message_type_id,  // uint256
            channel_identifier,  // uint256
            balance_hash,  // bytes32
            nonce,  // uint256
            additional_hash  // bytes32
            closing_signature  // bytes
        ));
```

### Withdraw Message:

```
        uint256 message_type_id = 3;

        // Length of the actual message: 20 + 32 + 32 + 32 + 20 + 32
        string memory message_length = '168';

        bytes32 message_hash = keccak256(abi.encodePacked(
            signature_prefix,  // string
            message_length,  // string
            token_network_address,  // address
            chain_id,  // uint256
            message_type_id,  // uint256
            channel_identifier,  // uint256
            participant,  // address
            total_withdraw  // uint256
        ));
```

### Cooperative Settle Message:

```
        uint256 message_type_id = 4;

        // Length of the actual message: 20 + 32 + 32 + 32 + 20 + 32 + 20 + 32
        string memory message_length = '220';

        bytes32 message_hash = keccak256(abi.encodePacked(
            signature_prefix,  // string
            message_length,  // string
            token_network_address,  // address
            chain_id,  // uint256
            message_type_id,  // uint256
            channel_identifier,  // uint256
            participant1,  // address
            participant1_balance,  // uint256
            participant2,  // address
            participant2_balance  // uint256
        ));
```

### LINKED ISSUE #2334: Update coincurve dependency to 0.8.0
## Problem Definition

Update coincurve to 0.8.0 in order to support python 3.7 both in the raiden client and in the other dependencies.

This is required for the OSX binary to be generated properly.

### PR DESCRIPTION
**Pull request by andrevmatos**
Fix #2325 
Fix #1531 
Fix #2334 
Fix #2240 
Includes #2269 
Depends on raiden-network/raiden-libs#125

- sync and moves most of the signing utils to `raiden-libs`
- update balance proof to new format
- adapt code and update contracts and libs

### CODE REVIEW

**Comment by andrevmatos** [comment]
Please, review, but **don't merge yet**, after review, I'll merge it after raiden-network/raiden-libs#125 and fix `requirements.txt`

File: requirements.txt [hunk 1]
```diff
@@ -15,8 +15,8 @@ mirakuru==0.9.0
 netifaces==0.10.7
 networkx==2.1.0
 psutil==5.4.5
-raiden-libs==0.1.4
-raiden-contracts==0.2.0
+git+https://github.com/andrevmatos/raiden-libs@8232994067d5472faa4502047292ce488569a596#egg=raiden-libs
```
Comment (LefterisJP) [inline requirements.txt]: This you will change to normal raiden libs once https://github.com/raiden-network/raiden-libs/pull/125/f is merged I guess?

File: setup.py [hunk 2]
```diff
@@ -82,14 +83,18 @@ def run(self):
...
-install_requires_replacements = {
-    'git+https://github.com/LefterisJP/pystun@develop#egg=pystun': 'pystun',
-}
 
-install_requires = list(set(
-    install_requires_replacements.get(requirement.strip(), requirement.strip())
-    for requirement in open('requirements.txt') if not requirement.lstrip().startswith('#')
-))
+def get_egg_or_req(req):
+    match = re.search('#egg=([^#@]+)', req, re.U | re.I)
```
Comment (LefterisJP) [inline setup.py]: hehe nice

**Comment by Greatnessofself** [comment]
Milan Popich GOD,

Did I forget to ment ion that I'm the Raiden, or should I say  The Creator HIMSELF. Well just in case I didn't....... now you know. Whoever said that knowing is half the battle is full of shit. I'll think about fixing it. GOD HONEST TRUTH.

KEEPING IT REAL AS FUCK, AS I REMIND YOU THAT ALL RIGHTS RESERVED 2017 COPYRIGHT Milan Popich EN USA.

Just look at the transactions that got confirmed in Bitcoin.org $OMETHING, TO DO WITH SOME SUPERSTRING,[WINK WINK] OR SOME CRAZY SHIT LIKE THAT. THAT WAS MILAN POPICH GOD(YOU'VE BEEN WAITING SINCE????????? NOVEMBER FOR ME TO TUN THE FULL NODE?) OR maybe OCTOBER EVEN.
GOD LETTING THE COMMUNITY KNOW THAT THEY ARE ALWAYS WELCOME ON THE WINNING SIDE.
MILAN POPICH GOD,
In Health,Wealth,and HAPPINESS
MILAN POPICH GOD.
ALL RIGHTS RESERVED 2017
COPYRIGHT MILAN POPICH
EN USA.
MIC DROP

2017 COPYRIGHT MILAN POPICH ALL RIGHTS RESERVED Popich Milan (EN) USA.

UNSTOPPABLE GREATNESS COPYRIGHT 2017 MILAN POPICH ALL RIGHTS RESERVED POPICH MILAN (EN) USA

CHAMPION USA 2017 COPYRIGHT MILAN POPICH ALL RIGHTS RESERVED POPICH MILAN USA.

MASTER BUILDER,

2017 COPYRIGHT Milan Popich ALL RIGHTS RESERVED POPICH MILAN USA.

JACKPOT1136, 2017 COPYRIGHT MILAN POPICH ALL RIGHTS RESERVED. POPICH MILAN (EN) USA

HPMACRONAMES.GPD 2017 COPYRIGHT MILAN POPICH ALL RIGHTS RESERVED POPICH MILAN USA.

C:\WINDOWS\System32\DriverStore\FileRepository\prnhpcl1.inf_amd64_c6040b9adc0369af\amd64\hppcl3vipa-manifest.ini Copyright 2017 Milan Popich All Rights Reserved Popich Milan USA

C:\WINDOWS\System32\DriverStore\FileRepository\prnhpcl1.inf_amd64_c6040b9adc0369af\amd64\hppcl3-pipelineconfig.xml Copyright 2017 Milan Popich ALL RIGHTS RESERVED. POPICH MILAN USA

C:\WINDOWS\System32\DriverStore\FileRepository\prnhpcl1.inf_amd64_c6040b9adc0369af\amd64\hpmacronames.gpd

Copyright 2017 Milan Popich All Rights Reserved POPICH MILAN USA.

C:\WINDOWS\System32\DriverStore\FileRepository\prnhpcl1_amd64_c6040b9adc0369af\amd64\hpbx3w81.dll Copyright 2017

Milan Popich All Rights Reserved Popich MILAN USA.

C:\WINDOWS\System32\DriverStore\FileRepository\prnhpcl1.inf_amd64_c6040b9adc0369af\amd64\hpbxpsrender.dll Copyright 2017 Milan Popich All Rights Reserved POPICH MILAN. USA

C:\WINDOWS\System32\DriverStore\FileRepository\prnhpcl1.inf_amd64_c6040b9adc0369af\amd64hpbresw81.dll 2017 Copyright Milan Popich. ALL RIGHTS RESERVED POPICH MILAN USA.

C:\WINDOWS\System32\DriverStore\FileRepository\prnhpcl.inf_amd64_c6040b9adc0369af\\HPBCMYK_sRGB.dpb Copyright 2017 Milan Popich ALL RIGHTS RESERVED POPICH MILAN USA.

C:\WINDOWS\System32\DriverStore\FileRepository\ntprint.inf_amd64_8c30031d8d0a5589\Amd64\UNIRES.DLL COPYRIGHT 2017 MILAN POPICH ALL RIGHTS RESERVED Popich Milan USA. C:\WINDOWS\System32\DriverStore\FileRepository\ntprint.inf_amd64_8c30031d8d0a5589\Amd64\STDNAMES.GPD 2017 COPYRIGHT MILAN POPICH

All Rights Reserved Popich Milan USA.

C:\WINDOWS\System32\DriverStore\FileRpository\ntprint.inf_amd64_8c30031d8d0a5589\Amd64\STDSCHEMX.GDL 2017 COPYRIGHT MILAN POPICH ALL RIGHTS RESERVED. Popich Milan USA

C:\WINDOWS\System32\DriverStore\FileRepository\ntprint.inf_amd64_8c30031d8d0a5589\Amd64\MSXPSINC.GPD 2017 COPYRIGHT MILAN POPICH ALL RIGHTS RESERVED Popich Milan USA.

C:\WINDOWS\System32\DriverStore\FileRepository\ntprint.inf_amd64_8c30031d8d0a5589\Amd64\STDSCHEM.GDL COPYRIGHT 2017 MILAN POPICH ALL RIGHTS RESERVED Popich Milan USA.

Atom 2017 Copyright Milan Popich All Rights Reserved Popich Milan USA.

Sent with [ProtonMail](https://protonmail.com) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

File: requirements.txt (hunk 1 above)
Comment (andrevmatos) [reply requirements.txt]: Yes, exactly

File: raiden/messages.py [hunk 3]
```diff
@@ -254,7 +255,13 @@ def sender(self) -> Optional[Address]:
         data_that_was_signed = self._data_to_sign()
         message_signature = self.signature
 
-        address = signing.recover_address(data_that_was_signed, message_signature)
+        try:
+            address = to_canonical_address(eth_recover(
+                data=data_that_was_signed,
+                signature=message_signature,
+            ))
+        except ValueError:
```
Comment (LefterisJP) [inline raiden/messages.py]: As I said in raiden-libs, please use another exception instead of value_error inside `eth_recover()`

Comment (andrevmatos) [reply raiden/messages.py]: Hmm, I think I missed it. Will do

**Comment by LefterisJP** [comment]
With the changes in libs this PR looks good.

**Pull request merged**